### DISCUSSION TITLE
Can I derive a TypedDict from a class?

### DISCUSSION THREAD

**Discussion by vadimcn** [microsoft/pyright#1591] [Q&A]
Is it possible to express the concept of "a TypedDict containing the same fields as \<some other class\>" to Pyright?  
For example:
```python
@dataclass
class Params:
    a: int = 42
    ...
def fucntion(**kwargs: AsTypedDict[Params]):
    ...
```
(I can't use a straight TypedDict, because they don't support defaults.  Besides, it's be nice to be able to use .field syntax for such objects, i.e. `params.a = 43`, rather than `params['a'] = 43`).

**Comment by erictraut** [comment]
No, there's nothing in the Python type standards that enable this. A TypedDict has very specific semantics as specified in [PEP 589](https://www.python.org/dev/peps/pep-0589/). If you simply "cast" a dataclass to a TypedDict, those semantic rules will be violated. For example, TypedDict doesn't allow the use of member access expressions (e.g. `params.a`) whereas dataclass does.

If you want to create a TypedDict from a class, you'd need to programmatically (at runtime) transform the dataclass object into a new TypedDict object.

```python
@dataclass
class Params:
    a: int = 42

class ParamsDict(TypedDict):
    a: int

def convert_params_to_dict(val: Params) -> ParamsDict:
    return {"a": val.a}
```

There's also currently no way to specify that a `**kwargs` parameter accepts values within a TypedDict because the notation `**kwargs: T` means that every value in the `kwargs` dictionary is of type `T`.

**Reply by vadimcn** [reply]
> There's also currently no way to specify that a **kwargs parameter accepts values within a TypedDict because the notation **kwargs: T means that every value in the kwargs dictionary is of type T.

:cry:

Still, I'm curious:  Is it possible to implement `AsTypedDict` in user code, or would this need to be a "magic type" in Pyright?

**Reply by erictraut** [reply]
I'm not sure what you mean by this. As I said above, you cannot simply "cast" a dataclass to a TypedDict. They have different semantics. So you would need to write a function that allocates a new dictionary object and copies the data out of the dataclass into the dictionary. If you want a type checker to understand the "shape" of that dictionary object (i.e. its allowed keys and corresponding value types), you would need to define that statically in the form of a `TypedDict` subclass as I showed above.
